### ISSUE TITLE
Gradle 5.0 executes tasks out of the order established by taskGraph

### ISSUE THREAD

**Issue by jbytecoder** [gradle/gradle#8082]
When I use the finalizedBy dependency between tasks. Gradle seems to evaluate the tasks in order different that the evaluated taskGraph suggests.  Details are self explanatory in the Steps to reproduce section

### Expected Behavior
Based on my understanding of finalizedBy dependency I expect the execution order to be:
[task ':doA', task ':cleanA', task ':doB', task ':cleanB', task ':doC', task ':cleanC', task ':doD', task ':cleanD', task ':rootTask'].
This should produce the following output:
`> Task :doA
Doing doA

> Task :cleanA
Doing cleanA

> Task :doB
Doing doB

> Task :cleanB
Doing cleanB

> Task :doC
Doing doC

> Task :cleanC
Doing cleanC

> Task :doD
Doing doD

> Task :cleanD
Doing cleanD

> Task :rootTask
`

### Current Behavior
When I run the build with "./gradlew -i rootTask" i can see the task graph it is as follows:
Tasks to be executed: [task ':doA', task ':cleanA', task ':doB', task ':cleanB', task ':doC', task ':cleanC', task ':doD', task ':cleanD', task ':rootTask']

But the output is quite diffrent from expectations:
`> Task :doA
Doing doA

> Task :doB
Doing doB

> Task :doC
Doing doC

> Task :doD
Doing doD

> Task :rootTask UP-TO-DATE

> Task :cleanA
Doing cleanA

> Task :cleanB
Doing cleanB

> Task :cleanC
Doing cleanC

> Task :cleanD
Doing cleanD
`

### Context
I discovered the problem while modifying my existing build. 

 I have a project that has it's sources generated from wsdl documents using wsimport utility.  But the project has also some common code at src/main/java.  I wan't to be able to "override" the code generated by wsimport with code from main source set. 

I am doing it the folowing way: Generate the sources using wsimport - scan the result looking for classes that already exists in main sourceset and remove them.

Up until now I used the the doLast hook in order to clean the results of wsimport, but i am also using the buildSrc helper project and each time I modify something inside it, gradle considers all the doLast hooks as dirty and causes rebuilds. I decided to move the build to using the proper Delete task and finalizedBy dependency to achieve the same result. Then I hit the bump.

### Steps to Reproduce 
In order to ilustrate this issue i have prepared a short build script: 
```
def root = tasks.register("rootTask") {
	def out = file("$buildDir/magic.txt")
	["A","B","C","D"].each {
		inputs.dir("$buildDir/dir$it")
	}
	outputs.file(out)

	doLast {
		println "done $name"
	}
}

["A","B","C","D"].each { srv ->
	def fin = tasks.register("clean$srv",Delete) {
		delete fileTree("$buildDir/dir$srv").filter {
			it.name == "out2.txt"
		}
		doLast {
			println "Doing $name"
		}
	}
	def run = tasks.register("do$srv") {
		inputs.file("build.gradle")
		outputs.dir("$buildDir/dir$srv")
		
		finalizedBy fin

		doFirst {
			["out1.txt","out2.txt"].each { f ->
			  file("$buildDir/dir$srv/$f").withWriter {
				it.writeLine "Content of $name for $f"
			  }	
			}
		}
		doLast {
			println "Doing $name"
		}
	}
	root.configure {
		dependsOn run
	}
}
```
This build requires no plugins or no other stuff - i did not include a complete archive because testing this issue is as simple as running gradle wrapper & pasting this as build.gradle

The idea behind this build is as follows: There exists a task (rootTask) that needs to operate on
output of 4 other tasks. The problem is that the output of those tasks has to be filtered - some files need to be removed. This is what the "clean" tasks are for it is imperative for the build that both "do"
tasks and clean tasks execute before the rootTask.   
I belive that the issue has to do with the up-to-date cheks and input caching. I think so because if we comment the inputs / outptus declarations in above build script the whole thing executes in expected order

### Your Environment
I belive in this case build scan is not necessary as anyone can simply copy the build script and see what is wrong

**Comment by wolfs** [comment]
@jbytecoder Thank you for reporting.
What was the last Gradle version where the tasks where executed as you expected? I guess you have the same problems with Gradle 5.0 and 5.1. Does 4.9 work?

**Comment by big-guy** [comment]
@jbytecoder I think you've figured it out.  There's no guarantee that a finalizer for one task will run before a task that depends on it. It used to be that's the way it worked out, but it hasn't been that way for awhile.

> It would be great if there was a way to tell gradle: always execute A and B together and then if I X.dependOn ':A' then before X executes A and B get executed ( their execution order should be decided between them according to dependOn and finalziedBy ).

This is a custom task :)

> When it comes to JavaExec / Exec tasks this becomes troublesome as you can't resue thoose tasks and have to reinvent the execution process plumbing that is already there.

There's a `project.javaexec` and `project.exec` that can be used as an imperative version of `JavaExec` and `Exec`.

I think you're reaching the end of where an ad-hoc task is manageable for the task at hand. There are some wsimport plugins on the plugin portal. 
This one looks like it could be good: https://plugins.gradle.org/plugin/uk.co.boothen.gradle.wsimport

**Comment by jbytecoder** [comment]
@big-guy Thank you for help with the second issue, I missed those helpfull little functions on project.  Looks like this will solve my problem.

Back to issue at hand. I thought that the response would be that there is no guarantee given for exact time on finalizedBy execution.  To be honest I would accepted this explanation if the resulting taskGraph also reflected it.  I decided to report it because the resulting task graph looks how I expect it to look while execution differs from this. 

I get the same result when I access gradle.taskGraph.allTasks - the order of tasks is as I expect it.
I think that this issue should not be closed, because documentation states that allTasks returns tasks included in execution plan: *in order* they will be executed. Clearly this is not the case in this build and the issue is about task ordering.  I focused on finalizedBy because I think it might be the cause

I would put It this way.  I always thought that after projects are evaluated, grade constructs taskGraph considering task dependencies( dependsOn,finalizedBy, and mustRunAfter ), and the the execution phase is guided by this taskGraph and allTasks specifically denotes the order in which tasks are executed.  

If the ```gradle.taskGraph.allTasks``` is not definite and gradle is allowed to reorder tasks in execution phase or simply execute them in different order, then there is no bug - maybe the documentation should be fixed.

If the documentation is correct then there is a bug.  It might be not affecting anything critical right know, it might not be urgent,  but should be checked
